### PR TITLE
Feature database and table count limitation issue56347

### PR DESCRIPTION
**Pull request by XuJia0210**
issue: https://github.com/ClickHouse/ClickHouse/issues/56347
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add server settings `max_table_num_to_throw` and `max_database_num_to_throw` to limit the number of databases or tables on `CREATE` queries.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [x]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 5df006098bd7c61b28d5502b50b8cabb917993e2 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64781/5df006098bd7c61b28d5502b50b8cabb917993e2/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 1]
```diff
@@ -138,6 +145,18 @@ BlockIO InterpreterCreateQuery::createDatabase(ASTCreateQuery & create)
             throw Exception(ErrorCodes::DATABASE_ALREADY_EXISTS, "Database {} already exists.", database_name);
     }
 
+    if (auto max_db = getContext()->getGlobalContext()->getServerSettings().max_database_num_to_throw; max_db > 0)
+    {
+        size_t db_count = DatabaseCatalog::instance().getDatabases().size();
+        // there's an invisible system database _temporary_and_external_tables, so we need to subtract 1
+        if (db_count > 0)
```
Comment (davenger) [inline src/Interpreters/InterpreterCreateQuery.cpp:152]: I think it makes sense to check that this hidden system DB is actually present  `&& DatabaseCatalog::instance().isDatabaseExist(DatabaseCatalog::TEMPORARY_DATABASE)`

File: tests/queries/0_stateless/03165_database_and_table_count_limitation_test.sh [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+# remove --database=$CLICKHOUSE_DATABASE from $CLICKHOUSE_CLIENT
+CLICKHOUSE_CLIENT=$(echo $CLICKHOUSE_CLIENT | sed "s/--database=$CLICKHOUSE_DATABASE//")
+
+# Create a temporary directory
+TEMP_DIR=$(mktemp -d "$CUR_DIR/$(basename "$BASE").XXXXXX")
+
+# Generate config.xml
+CONFIG_FILE="$TEMP_DIR/config.xml"
+cat > "$CONFIG_FILE" <<EOL
+<yandex>
```
Comment (hanfei1991) [inline tests/queries/0_stateless/03165_database_and_table_count_limitation_test.sh:22]: 1. We have left yandex :)
2. Don't do this in stateless test :(  we write integration tests to test special config

File: tests/queries/0_stateless/03165_database_and_table_count_limitation_sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
+create database db1;
```
Comment (hanfei1991) [inline tests/queries/0_stateless/03165_database_and_table_count_limitation_sql:1]: Don't do this .... we only have `.sql` and `.reference` in `0_stateless` directory

**Comment by hanfei1991** [comment]
I write an integration test to replace original one. Feel free to amend/revert if you do not agree :)

I have another question, why we don't care about system tables for tables limit, but have to care `system` and other default databases for limit? We can make the behaviour unified: only care about database/table which users create.

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 4]
```diff
@@ -138,6 +145,18 @@ BlockIO InterpreterCreateQuery::createDatabase(ASTCreateQuery & create)
             throw Exception(ErrorCodes::DATABASE_ALREADY_EXISTS, "Database {} already exists.", database_name);
     }
 
+    if (auto max_db = getContext()->getGlobalContext()->getServerSettings().max_database_num_to_throw; max_db > 0)
```
Comment (hanfei1991) [inline src/Interpreters/InterpreterCreateQuery.cpp:148]: We can write two lines, don't have to be tooo brief

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 5]
```diff
@@ -1537,6 +1556,16 @@ bool InterpreterCreateQuery::doCreateTable(ASTCreateQuery & create,
         }
     }
 
+    if (UInt64 max_table = getContext()->getGlobalContext()->getServerSettings().max_table_num_to_throw; max_table > 0)
```
Comment (hanfei1991) [inline src/Interpreters/InterpreterCreateQuery.cpp:1559]: ditto

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 6]
```diff
@@ -1537,6 +1556,16 @@ bool InterpreterCreateQuery::doCreateTable(ASTCreateQuery & create,
...
+    {
+        UInt64 table_count = CurrentMetrics::get(CurrentMetrics::AttachedTable);
+        if (table_count >= max_table)
+            throw Exception(ErrorCodes::TOO_MANY_TABLES,
+                            "Too many tables in the system. Current is {}, limit is {}. "
```
Comment (hanfei1991) [inline src/Interpreters/InterpreterCreateQuery.cpp:1564]: How about using one sentence:
`Too many tables in ClickHouse. The limit (setting 'max_table_num_to_throw') is set to {}`

Same for databases.

Besides, do write period at the end.

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**